homeassistant/components/amcrest/camera.py

```diff
@@ -170,18 +208,183 @@ def turn_on(self):
+        """Call the job and set camera color mode."""
+        return self.hass.async_add_job(self._set_color_bw, cbw)
+
+    @callback
+    def async_start_tour(self):
+        """Call the job and start camera tour."""
+        return self.hass.async_add_job(self._start_tour, True)
+
+    @callback
+    def async_stop_tour(self):
+        """Call the job and stop camera tour."""
+        return self.hass.async_add_job(self._start_tour, False)
+
+    # Methods to send commands to Amcrest camera and handle errors
 
     def _enable_video_stream(self, enable):
         """Enable or disable camera video stream."""
         from amcrest import AmcrestError
 
+        # Given the way the camera's state is determined by
+        # is_streaming and is_recording, we can't leave
+        # recording on if video stream is being turned off.
+        if self.is_recording and not enable:
+            self._enable_recording(False)
         try:
-            self._camera.video_enabled = enable
+            self._api.video_enabled = enable
         except AmcrestError as error:
             _LOGGER.error(
                 'Could not %s %s camera video stream due to error: %s',
                 'enable' if enable else 'disable', self.name, error)
         else:
             self.is_streaming = enable
             self.schedule_update_ha_state()
+
+    def _enable_recording(self, enable):
+        """Turn recording on or off."""
+        from amcrest import AmcrestError
+
+        # Given the way the camera's state is determined by
+        # is_streaming and is_recording, we can't leave
+        # video stream off if recording is being turned on.
+        if not self.is_streaming and enable:
+            self._enable_video_stream(True)
+        rec_mode = {'Automatic': 0, 'Manual': 1}
+        try:
+            self._api.record_mode = rec_mode[
+                'Manual' if enable else 'Automatic']
+        except AmcrestError as error:
+            _LOGGER.error(
+                'Could not %s %s camera recording due to error: %s',
+                'enable' if enable else 'disable', self.name, error)
+        else:
+            self._is_recording = enable
+            self.schedule_update_ha_state()
+
+    def _enable_motion_detection(self, enable):
+        """Enable or disable motion detection."""
+        from amcrest import AmcrestError
+
+        try:
+            self._api.motion_detection = str(enable).lower()
+        except AmcrestError as error:
+            _LOGGER.error(
+                'Could not %s %s camera motion detection due to error: %s',
+                'enable' if enable else 'disable', self.name, error)
+        else:
+            self._motion_detection_enabled = enable
+            self.schedule_update_ha_state()
+
+    def _enable_audio(self, enable):
+        """Enable or disable audio stream."""
+        from amcrest import AmcrestError
+
+        try:
+            self._api.audio_enabled = enable
```

Comment:
This takes a boolean but not `_api.motion_detection`?

Reply:
The camera API is bad. That became obvious in previous PRs. It uses properties and setters when they should be methods. I will hopefully change that some day. But for now, `self._api.audio_enabled` is a setter that will actually send a command to the camera via the network, and the value sent depends on the value being written to the setter, in this case `enable`. But, yes, I agree, given how properties and setters _should_ be used, this is confusing, but I can't fix it without making a huge breaking change to the amcrest pypi.org package.